pkg-r/R/contents_shinychat.R

```
@@ -1,14 +1,296 @@
+    intent = content@request@arguments$.intent,
+    expanded = if (isTRUE(content@extra$display$open)) NA
+  )
+
+  icon_deps <- NULL
+  tool <- content@request@tool
+
+  # Let tool results override global tool properties
+  if (!is.null(content@extra$display$title)) {
+    props$title <- content@extra$display$title
+  }
+  if (!is.null(content@extra$display$icon)) {
+    props$icon <- content@extra$display$icon
+    icon_deps <- htmltools::findDependencies(props$icon)
+  }
+
+  if (!is.null(tool)) {
+    # Format fails if tool is not present (ellmer v0.3.0, tidyverse/ellmer#691)
+    props$request_call <- format(content@request, show = "call")
+
+    props$title <- props$title %||% tool@annotations$title
+    props$icon <- props$icon %||% tool@annotations$icon
+    icon_deps <- icon_deps %||% htmltools::findDependencies(props$icon)
+  } else {
+    props$request_call <- jsonlite::toJSON(
+      list(
+        id = content@request@id,
+        name = content@request@name,
+        arguments = content@request@arguments
+      ),
+      auto_unbox = TRUE,
+      pretty = 2
+    )
+  }
+
+  props <- list2(!!!props, !!!tool_result_display(content))
+  names(props) <- gsub("_", "-", names(props))
+
+  # Get dependencies for the value because we store its HTML as an attribute
+  # where `processDeps()` won't find it.
+  deps <- htmltools::findDependencies(props$value)
+
+  htmltools::tagList(
+    htmltools::tag("shiny-tool-result", props),
+    deps,
+    icon_deps,
+    chat_deps()
+  )
+}
+
+ensure_content_display <- function(content) {
+  display <- content@extra$display
+
+  if (is.null(display) || opt_shinychat_tool_display() == "basic") {
+    return(list())
+  }
+
+  invalid_display_fmt <- "Invalid {.code @extra$display} format for {.code ContentToolResult} from {.fn {content@request@name}} (call id: {content@request@id})."
+
+  if (
+    inherits(display, c("html", "shiny.tag", "shiny.tag.list", "htmlwidgets"))
+  ) {
```

FWIW, if `display` is a character string or a bare list, it currently gets silently dropped

---

No, the check in the lines just after this catch the case where `display` is a string. And lists aren't dropped.

``` r
library(ellmer)
pkgload::load_all()
#> ℹ Loading shinychat

the_tool <- tool(
  \() "output",
  name = "tool_call",
  description = "A tool that returns output"
)

request <- ContentToolRequest(
  id = "abc123",
  name = "tool_call",
  tool = the_tool
)

ContentToolResult(
  value = "output",
  extra = list(display = "bad!"),
  request = request
) |>
  contents_shinychat() |>
  format()
#> Warning: Invalid `@extra$display` format for `ContentToolResult` from `tool_call()`
#> (call id: abc123).
#> ✖ Expected a list with fields `html`, `markdown`, `text`, `show_request`,
#>   `open`, `title`, or `icon`, not a string.
#> [1] "<shiny-tool-result request-id=\"abc123\" request-call=\"tool_call()\" name=\"tool_call\" status=\"success\" show-request value=\"output\" value-type=\"code\"></shiny-tool-result>"

ContentToolResult(
  value = "output",
  extra = list(display = list(markdown = "this works fine")),
  request = request
) |>
  contents_shinychat() |>
  format()
#> [1] "<shiny-tool-result request-id=\"abc123\" request-call=\"tool_call()\" name=\"tool_call\" status=\"success\" show-request value=\"this works fine\" value-type=\"markdown\"></shiny-tool-result>"
```

---

Ahh right, my bad.